File: packages/wasmvm/wasmclient/src/isc/waspclient.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

pub use crate::offledgerrequest::*;
pub use crate::receipt::*;
pub use std::time::*;
pub use wasmlib::*;

const DEFAULT_OPTIMISTIC_READ_TIMEOUT: Duration = Duration::from_millis(1100);

pub struct WaspClient {
    base_url: String,
    token: String,
}

impl WaspClient {
    pub fn new(base_url: &str) -> WaspClient {
        return WaspClient {
            base_url: base_url.to_string(),
            token: String::from(""),
        };
    }
    pub fn call_view_by_hname(
        &self,
        chain_id: &ScChainID,
        contract_hname: &ScHname,
        function_hname: &ScHname,
        args: ScDict,
        optimistic_read_timeout: Option<Duration>,
    ) -> Result<(), String> {
        let deadline = match optimistic_read_timeout {
            Some(duration) => duration,
            None => DEFAULT_OPTIMISTIC_READ_TIMEOUT,
        };

        let url = format!(
            "/chain/{}/contract/{}/callviewbyhname/{}",
            chain_id.to_string(),
            contract_hname.to_string(),
            function_hname.to_string()
        );
        let client = reqwest::blocking::Client::builder()
            .timeout(deadline)
            .build()
            .unwrap();
        let _ = client.post(url).body(args.to_bytes()).send();
        Ok(())
    }
    pub fn post_offledger_request(
        &self,
        chain_id: &ScChainID,
        req: &OffLedgerRequestData,
    ) -> Result<(), String> {
        let url = format!("/chain/{}/request", chain_id.to_string(),);
        let client = reqwest::blocking::Client::new();
        let _ = client.post(url).body(req.to_bytes()).send();
        Ok(())
    }
    pub fn wait_until_request_processed(
        &self,
        chain_id: &ScChainID,
        req_id: &ScRequestID,
        timeout: Duration,
    ) -> Result<(), String> {
        let url = format!(
            "/chain/{}/request/{}/wait",
            chain_id.to_string(),
            req_id.to_string()
        );
        let client = reqwest::blocking::Client::builder()
            .timeout(timeout)
            .build()
            .unwrap();
        let _ = client.get(url).send();
        Ok(())
    }
}

File: packages/wasmvm/wasmclient/ts/wasmclient/lib/isc/waspclient.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import * as wasmlib from 'wasmlib';
import {SyncRequestClient} from './ts-sync-request';
import {OffLedgerRequest} from './offledgerrequest';
import {APICallViewRequest, APIOffLedgerRequest, Codec, JsonReq, JsonResp} from './codec';

export type Error = string | null;

export class WaspClient {
    baseURL: string;

    public constructor(baseURL: string) {
        this.baseURL = baseURL;
    }

    public callViewByHname(chainID: wasmlib.ScChainID, hContract: wasmlib.ScHname, hFunction: wasmlib.ScHname, args: Uint8Array): [Uint8Array, Error] {
        const url = this.baseURL + '/requests/callview';
        const req = new SyncRequestClient();
        req.addHeader('Content-Type', 'application/json');

        const callViewRequest: APICallViewRequest = {
            contractHName: hContract.toString(),
            functionHName: hFunction.toString(),
            chainId: chainID.toString(),
            arguments: Codec.jsonEncode(args),
        };

        try {
            const resp = req.post<APICallViewRequest, JsonResp>(url, callViewRequest);
            const result = Codec.jsonDecode(resp);
            return [result, null];
        } catch (error) {
            let message;
            if (error instanceof Error) message = error.message;
            else message = String(error);
            return [new Uint8Array(0), message];
        }
    }

    public postOffLedgerRequest(chainID: wasmlib.ScChainID, signed: OffLedgerRequest): Error {
        const url = this.baseURL + '/requests/offledger';
        const req = new SyncRequestClient();
        req.addHeader('Content-Type', 'application/json');

        const offLedgerRequest: APIOffLedgerRequest = {
            chainId: chainID.toString(),
            request: wasmlib.hexEncode(signed.bytes()),
        };

        try {
            req.post(url, offLedgerRequest);
            return null;
        } catch (error) {
            let message;
            if (error instanceof Error) message = error.message;
            else message = String(error);
            return message;
        }
    }

    public waitUntilRequestProcessed(chainID: wasmlib.ScChainID, reqID: wasmlib.ScRequestID, timeout: u32): Error {
        //TODO Timeout of the wait can be set with `/wait?timeoutSeconds=`. Max seconds are 60secs.
        const url = this.baseURL + '/chains/' + chainID.toString() + '/requests/' + reqID.toString() + '/wait';
        const response = new SyncRequestClient().get(url);
        return null;
    }
}
